### ISSUE TITLE
Having a .moshrc (equivalent to .ssh/config) file to be able to specify per-host configs

### ISSUE THREAD

**Issue by lemoinem** [mobile-shell/mosh#431]
Hi,

I think this would go a long way improving the usability of mosh to be able to have a .ssh/config equivalent.

Unfortunately, ssh complains about unknown directives in .ssh/config so we will need a new file (I also think it's good practice to do so anyway).

The format could be simple (.ssh/config sounds a pretty good example: easy to write, easy to parse).

Useful configs to include: `hostname`, `port`, `mosh-server`, eventually `ssh` and `command` (in some weird case). Later: enable/disable OOB data (#423) on a per-host basis.
Some of the configs such as `hostname`, `agent forwarding`, `X11 forwarding`, etc... could actually be extracted from .ssh/config directly.

I think it's much better than aliases because it's seamless and consistent. Plus the fact that mosh already seamlessly use .ssh/config makes it quite irritating to be able to use aliases only for specifying a couple of argument for a given host. If many hosts requires a custom config multiplication of aliases could also clutter the shell. Plus we always have to remember which hosts works out-of-the-box and which needs an alias. All this disappears with a `.mosh/config` config file.

Security-wise, since it's already done by ssh (and actually allows a much bigger control on the connection since it could change the Hostname we are connecting to, define a ProxyCommand, etc...), If the file is secure, I don't see a very big security issue here.

**Comment by infinity0** [comment]
Instead of having a separate `.moshrc`, it would be easier for users, to just put these extra options in `.ssh/config`?

~~~~
Host xxx.com.yyy
HostName xxx.com
User uuu
MoshPort 6nnn
MoshPrediction always

**Comment by andersk** [comment]
@infinity0, nope, the unrecognized syntax would completely break `ssh` (and therefore also `mosh`). I could imagine putting them inside magic comments…but not for two seconds longer than I have to.

**Comment by emtiu** [comment]
Any consensus on this in sight? I'm also very much interested in server-side custom port(range)s. Don't mind if it's systemwide (as in /etc/moshports) or per-user (as in ~/.moshports).

**Comment by Toub** [comment]
Also interested to have a client-side or server-side configuration for udp port range.

**Comment by rgruesbeck** [comment]
+1 for this

**Comment by Myke** [comment]
Another vote for server-side config; with multiple servers behind NAT, ssh-config nicely solves the port changes, but there's no elegant way to teach mosh that it's only getting specific ports forwarded to it.

**Comment by yookoala** [comment]
+1

There is no way to config custom mosh-server udp port range now.
Would be really handy if a config file is available.

**Comment by tjanson** [comment]
Please refrain from "+1-ing" issues, here or elsewhere. I’m sure you mean well, but it dilutes the discussion and essentially spams subscribers of the issue.

**Comment by skmlcd** [comment]
Great idea, but please avoid polluting home directory, refer to [XDG base directory spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html).

**Comment by dpsenner** [comment]
Is there a plan when this feature might be released?

**Comment by nepthar** [comment]
I know this is an old ticket, but I've noticed that it's still open. Could you provide some guidance as to whether this is still in the works? I have several machines running mosh behind NAT and it would be fantastic if I could configure them server side to use a specified port range.

**Comment by cyqsimon** [comment]
Bump. Since the server is already responsible for selecting a port, it makes a lot of sense for there to be a config option to pin this port choice.

I have several servers behind a common gateway, and currently I have to remember each of their UDP port mapping which is a pain in the arse. Hopefully this feature gets released soon.

**Comment by ioogithub** [comment]
Can we just have mosh parse the ports from ~/.ssh/config? I already have all my hosts setup with the ports the require but mosh seems to ignore this and simple tries to connect to the default ports. Is this a bug or a design limitation?

I can't see mosh being very useful if I need to remember and type in long string with UPD ports for every server.

**Comment by yookoala** [comment]
> Can we just have mosh parse the ports from ~/.ssh/config? I already have all my hosts setup with the ports the require but mosh seems to ignore this and simple tries to connect to the default ports. Is this a bug or a design limitation?
> 
> I can't see mosh being very useful if I need to remember and type in long string with UPD ports for every server.

mosh uses ssh to initiate it's own service. In this phase, mosh uses your `Port` setting in `~/.ssh/config` to establish ssh connection. But mosh and ssh are different services, they can't use the same set of ports or else they would conflict with each other.

**Comment by cyqsimon** [comment]
> Can we just have mosh parse the ports from ~/.ssh/config? I already have all my hosts setup with the ports the require but mosh seems to ignore this and simple tries to connect to the default ports. Is this a bug or a design limitation?
> 
> I can't see mosh being very useful if I need to remember and type in long string with UPD ports for every server.

Mosh does respect `.ssh/config`. If you've defined custom hosts in there, you can mosh into them using their aliases just as you would using SSH.

The difference is, mosh only uses the SSH profile during initial establishment of contact. Then mosh set up its own UDP server listening on another port (completely unrelated to SSH, and this is the port we want a config for), and tells the client to contact it over there. Kind of like how ATC handles switchovers in the aviation industry. It is obvious then why settings for this mosh UDP server should belong in its own config file. Hopefully this lets you understand how mosh works a little bit better.

**Comment by ryneches** [comment]
I just wish mosh would use a more of the directives within the aliases in `.ssh/config`, particularly `LocalForward`. That would be awesome.

**Comment by frederickjh** [comment]
@cyqsimon The work around that I use to set the mosh ports is to use a wrapper script to set the ports based on the hostname. I use  the [fish shell](https://fishshell.com/) and I use a fish shell function for this. You could also make a bash wrapper script but name it something different like, 'start-mosh'. Here is a sanitized version of the fish function:
```fish
# Defined in /home/username/.config/fish/functions/mosh.fish @ line 1
function mosh --description 'mosh wrapper to setup special command line option for hosts that we do not want to type or remember each time.'
    set sshhostname $argv[1]
    set -e argv[1]
    switch $sshhostname
        case example.com
	        set runoptions -p 60077 --port 60077:60086
        case example.org
	        set runoptions -p 60088 --port 60088:60096
        case '*'
     end
     set_color yellow
    echo -n "Now running: "
    set_color blue
    echo /usr/bin/mosh $runoptions $argv $sshhostname
    set_color normal
    echo ""
    /usr/bin/mosh $runoptions $argv $sshhostname
end
```
I hope that someone finds this workaround helpful.

**Comment by burakozhan** [comment]
### tl;dr
Need to enter port manually each time? -> RTFM ;)

### Long version:
I do have multiple servers worldwide at different operators. One of them does an external Firewalling outside of the usual linux VM firewall. So with a proper ssh config file, while for other servers I simply tpye `mosh d` to connect to delta-server in delaware USA, for the German server I painfully had to remember typing in `mosh g -p 64829` every time, because the golf-server needs a punch in the udp port. And on the firewall config I can not give a port-range, I have to manually set each port one-by-one on a pretty slow web-site that takes the configuration. Entering 1000 ports is not feasible (and I don't even know if possible) Yes, I rent the cheapest crappy vps I can find and yes, I name my servers alpha, bravo, charlie, etc. In my ssh config thety are listed as Host a, Host b, Host c, ...

From former info I have had from the command line output which reads: 
```(By default, mosh uses a UDP port between 60000 and 61000. The -p option selects a specific UDP port number.)```
I was _assuming_ mosh uses a _random_ port between 60000 and 61000. But that is _not_ the case. I uses a sequential port. It **always starts with 60001** for the first session. Only if there is a parallel session it needs 60002. 

**The FAQ** already states that:
```Mosh will use the first available UDP port, starting at 60001 and stopping at 60999. If you are only going to have a small handful of concurrent sessions on a server, then you can forward a smaller range of ports (e.g., 60000 to 60010).```

So the authors of mosh are a bit foggy around the edges there. Does it start at 60000 or 60001 and does it end with 61000 or 60999?

So for all that have still problems with this:
Open 60001 - 60005 or so (for the occasional hanging session) on your firewall for UDP and all will be fine. 
(60000 is not used, it starts with 60001)
Also; since the port does not listen, as long as you don't try to connect to the server, there is no permanent security hole on port 60001.

Still, having the port range configurable on the server side would be nice, just to match the firewall of the server for unknowing users, but I could solve my problem in another way.

**Comment by qtfkwk** [comment]
Just ran into this "problem" and thought it'd be a nice feature.

Saw prior comments about doing it server side, anyone know if that was ever implemented?

Hat tip to @frederickjh for the fish function https://github.com/mobile-shell/mosh/issues/431#issuecomment-1803228704

Wrote a bash script (more widely usable than @frederickjh's fish function) to sit in `~/bin/mosh` (or anywhere ahead of real mosh location in `$PATH`); it's pretty minimal but:

- Creates a template config file if it doesn't exist
- Shows usage with no args
- Passes through to "real mosh" if `-p` used or host not in config
- Errors if host is in config but no port
- Otherwise grabs the port for the host from the config and uses it

Note that it's still pretty basic (especially config parsing, etc)... so no guarantees/licenses/etc

`~/bin/mosh`:

```bash
#!/usr/bin/env bash

CONFIG="$HOME/.mosh/config"

mkdir -p "$(dirname "$CONFIG")"

if [ ! -e "$CONFIG" ]; then
  cat <<"EOF" >"$CONFIG"
# Host $HOST
# Port $PORT
EOF
  echo -e "Created config file \`$CONFIG\`\n"
fi

HOST="$1"

if [ -z "$HOST" ]; then
  echo Usage: "$(basename "$0")" HOST
  exit 1
fi

MOSH="$(which -a mosh |sed -n 2p)"

if echo "$@" |grep "\-p " >/dev/null; then
  # Command already has `-p`; use it
  exec "$MOSH" $@
elif grep "^Host $HOST" "$CONFIG" >/dev/null; then
  # Host in config; get and use the configured port
  PORT=$(sed -n "/^Host $HOST/,/^$/p" "$CONFIG" |grep ^Port |cut -d\  -f2)
  if [ -z "$PORT" ]; then
    echo "Host \`$HOST\` is in config file \`$CONFIG\` but there is no configured port"
    exit 2
  else
    exec "$MOSH" -p "$PORT" "$HOST"
  fi
else
  # Host not in config; use default port(s)
  exec "$MOSH" "$HOST"
fi
```

**Comment by erudenko** [comment]
While waiting for native support, here's a **zsh wrapper** approach that reads mosh-specific settings from a dedicated `~/.ssh/mosh` config file using the same format as SSH config:

### Config file: `~/.ssh/mosh`

```
Host my-server
  MoshServer /opt/homebrew/bin/mosh-server
  MoshCommand /Users/jack/.local/bin/tmux-login

Host another-server
  MoshServer /usr/local/bin/mosh-server
```

### Wrapper function: `~/.ssh/mosh.zsh`

```zsh
# mosh wrapper: reads MoshServer and MoshCommand from ~/.ssh/mosh
# Source this file in .zshrc: source ~/.ssh/mosh.zsh
mosh() {
    local mosh_server="" mosh_command="" host=""

    local skip_next=""
    for arg in "$@"; do
        if [[ -n "$skip_next" ]]; then
            skip_next=""
            continue
        fi
        case "$arg" in
            -p|-n|--ssh) skip_next=1 ;;
            -*) ;;
            *) host="$arg"; break ;;
        esac
    done

    if [[ -n "$host" && -f ~/.ssh/mosh ]]; then
        local in_host=0
        while IFS= read -r line; do
            if [[ "$line" =~ ^[[:space:]]*Host[[:space:]]+(.*) ]]; then
                [[ "${match[1]%% *}" == "$host" ]] && in_host=1 || in_host=0
                continue
            fi
            if (( in_host )); then
                [[ "$line" =~ ^[[:space:]]*MoshServer[[:space:]]+(.*[^[:space:]]) ]] && mosh_server="${match[1]}"
                [[ "$line" =~ ^[[:space:]]*MoshCommand[[:space:]]+(.*[^[:space:]]) ]] && mosh_command="${match[1]}"
            fi
        done < ~/.ssh/mosh
    fi

    local -a cmd=(command mosh)
    [[ -n "$mosh_server" ]] && cmd+=(--server="$mosh_server")
    cmd+=("$@")
    [[ -n "$mosh_command" ]] && cmd+=(-- "$mosh_command")

    "${cmd[@]}"
}
```

### Usage

```bash
# Add to .zshrc
source ~/.ssh/mosh.zsh

# Then just:
mosh my-server          # uses MoshServer + MoshCommand from config
mosh other-host         # no config match → plain passthrough
mosh -p 60001 my-server # extra flags are forwarded
```

**Notes:**
- Uses a separate `~/.ssh/mosh` file (not SSH config) because SSH rejects unknown directives
- Hosts without entries in the config file pass through to mosh unchanged
- Handles `-p` and `-n` flags that take separate value arguments
- Uses `~` expansion caveat: use absolute paths for `MoshCommand` since mosh-server runs it via `execvp` without shell expansion

**Comment by tuxick** [comment]
> Just ran into this "problem" and thought it'd be a nice feature.
> Wrote a bash script (more widely usable than [@frederickjh](https://github.com/frederickjh)'s fish function) to sit in `~/bin/mosh` (or anywhere ahead of real mosh location in `$PATH`); it's pretty minimal but:

Great, does the job.
